Signal/src/ViewControllers/AddContactShareToExistingContactViewController.swift

```diff
@@ -55,17 +55,29 @@ class AddContactShareToExistingContactViewController: ContactsPicker, ContactsPi
         navigationController.popViewController(animated: true)
     }
 
-    func contactsPicker(_: ContactsPicker, didSelectContact contact: Contact) {
+    func contactsPicker(_: ContactsPicker, didSelectContact oldContact: Contact) {
         Logger.debug("\(self.logTag) in \(#function)")
 
-        guard let mergedContact: CNContact = self.contactShare.cnContact(mergedWithExistingContact: contact) else {
-            owsFail("\(logTag) in \(#function) mergedContact was unexpectedly nil")
+        let contactsManager = Environment.current().contactsManager
+        guard let oldCNContact = contactsManager?.cnContact(withId: oldContact.cnContactId) else {
```

Comment:
Everywhere we are using `CNContact` becomes slightly more laborious, but this change dramatically reduces memory usages for users with large numbers of contacts.

Reply:
We've also recently added some fields in:

```
    public static let allowedContactKeys: [CNKeyDescriptor] = [
        CNContactFormatter.descriptorForRequiredKeys(for: .fullName),
        CNContactThumbnailImageDataKey as CNKeyDescriptor,
        CNContactPhoneNumbersKey as CNKeyDescriptor,
        CNContactEmailAddressesKey as CNKeyDescriptor,
        CNContactPostalAddressesKey as CNKeyDescriptor,
        CNContactViewController.descriptorForRequiredKeys(),
        CNContactVCardSerialization.descriptorForRequiredKeys()
    ]
```

I wonder if any of those could be trimmed out?

Reply:
Great question - I don't think so, though.

* `CNContactThumbnailImageDataKey` we recently learned that some `CNContacts` have images and some have thumbnails, some have both and some have neither.  To make a best effort to show an avatar, we need both keys.  I think the thumbnails are always smaller.
* We need the name, phone numbers, emails, addresses.
* `CNContactViewController.descriptorForRequiredKeys` we need this for contact editing.
* `CNContactVCardSerialization.descriptorForRequiredKeys` we need this for contact sharing. We don't do it yet, but my understanding is that this is currently a subset of the keys we already use so it should only serve to future-proof us.